### PR TITLE
Support Proxy

### PR DESCRIPTION
**Pull request by ioedeveloper**
closes https://github.com/ethereum/remix-project/issues/2414

### Testing
* Build and run support proxy PR
* Create an upgradeable smart contract using Openzeppelin wizard by checking the upgradeability UUPS option. https://docs.openzeppelin.com/contracts/4.x/wizard
* Copy and paste the generated code into remix-ide.
* Compile the smart contract and open the Udapp plugin.
* Select the deploy with proxy option

* And deploy
* The Udapp should have 2 instances to interact with the contract. The first is the original contract implementation and the second is the proxy deployment.

### CODE REVIEW

**Review by AlxKrispy — APPROVED** [review]

File: apps/remix-ide/src/app/udapp/run-tab.js [hunk 1]
```diff
@@ -149,4 +150,11 @@ export class RunTab extends ViewPlugin {
   readFile (fileName) {
     return this.call('fileManager', 'readFile', fileName)
   }
+
+  resolveContract (contractObject, address) {
```
Comment (yann300) [inline apps/remix-ide/src/app/udapp/run-tab.js:154]: could you rename it `resolveContractAndAddInstance`

File: apps/remix-ide/src/blockchain/blockchain.js [hunk 2]
```diff
@@ -127,13 +129,53 @@ export class Blockchain extends Plugin {
...
+      if (error) {
+        return statusCb(`creation of ${selectedContract.name} errored: ${error.message ? error.message : error}`)
+      }
 
       statusCb(`creation of ${selectedContract.name} pending...`)
       this.createContract(selectedContract, data, continueCb, promptCb, confirmationCb, finalCb)
     })
   }
 
+  async deployProxy (proxyData, implementationContractObject) {
```
Comment (yann300) [inline apps/remix-ide/src/blockchain/blockchain.js:141]: This'd be necessary to show a modal to explain what we are about to deploy

File: libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts [hunk 3]
```diff
@@ -1,26 +1,80 @@
...
 export class OpenZeppelinProxy extends Plugin {
   blockchain: any
+  kind: 'UUPS' | 'Transparent'
   constructor(blockchain) {
     super(proxyProfile)
     this.blockchain = blockchain
   }
 
-  async isConcerned(contractData: ContractData): Promise<boolean> {
+  async isConcerned(ast: ContractAST = {} as ContractAST) {
```
Comment (yann300) [inline libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts:19]: this should return true or false: `Promise<boolean>`

File: libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts [hunk 4]
```diff
@@ -1,26 +1,80 @@
...
+      return true
+    }
+    //
+    // else if transparent contract run check true/false
+    //
     return false
   }
 
-  async execute(contractData: ContractData, contractMetadata: any, compiledContracts: any) {
+  async getDeployOptions (contracts: ContractABI) {
```
Comment (yann300) [inline libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts:31]: this should return a map of inputs.

File: libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts [hunk 5]
```diff
@@ -1,26 +1,80 @@
...
+    if (this.kind === 'UUPS') {
```
Comment (yann300) [inline libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts:32]: if not

Comment (yann300) [inline libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts:32]: by default it should return an empty mapping.

File: libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts [hunk 6]
```diff
@@ -1,26 +1,80 @@
...
+      const inputs = {}
+    
+      Object.keys(contracts).map(name => {
+        const abi = contracts[name].abi
+        const initializeInput = abi.find(node => node.name === 'initialize')
```
Comment (yann300) [inline libs/remix-core-plugin/src/lib/openzeppelin-proxy.ts:37]: how are you sure you are targeting the righ contract? should `getDeployOptions` have a parameter like `contractName: string` ?

File: libs/remix-ui/run-tab/src/lib/run-tab.tsx [hunk 7]
```diff
@@ -134,7 +134,7 @@ export function RunTabUI (props: RunTabProps) {
 
   const handleToaster = () => {
     setFocusToaster('')
-    clearPopUp()
+    hideToaster()
```
Comment (yann300) [inline libs/remix-ui/run-tab/src/lib/run-tab.tsx:137]: no more `clearPopUp`?

**Comment by ryestew** [comment]
For the modal: 
**Title**
Current
ERC1967

**Proposed**
2 contracts will be deployed
or
Deployment of your contract AND an ERC1967 proxy contract

**Body**
Current
Confirm proxy deployment?

Proposed
Confirm you want to deploy 2 contracts: your implementation AND an ERC1967 proxy contract.

**Review by ryestew — APPROVED** [review]

**Pull request merged**